test-suite/piecewiseyieldcurve.cpp

```diff
@@ -1159,6 +1159,9 @@ void PiecewiseYieldCurveTest::testBadPreviousCurve() {
         { 50, Years,   0.01185 }
     };
 
+    //make this test run standalone
+    Settings::instance().evaluationDate() = Date(10, May, 2022);
```

Comment:
I'm not sure I understand what's happening here?

Reply:
The test case fails if you run only this case. A previous test case has to set the evaluation date for this test to pass. I do not understand the test case enough to answer why this is necessary.

Reply:
I see — unlike the default runner, the parallel one doesn't execute [this bit](https://github.com/lballabio/QuantLib/blob/master/test-suite/quantlibtestsuite.cpp#L315) which sets the evaluation date.  Is it possible to add it?

If not, I'd try moving up here the lines, a bit further down, that set the evaluation date to `today`.

Reply:
Actually the parallel test runner executes [this bit](https://github.com/lballabio/QuantLib/blob/master/test-suite/quantlibtestsuite.cpp#L315) in every instance but the date has to be past [this date](https://github.com/lballabio/QuantLib/blob/master/test-suite/piecewiseyieldcurve.cpp#L1170). Unfortunately this is not true for QuantLib's default date, the 16.09.2015 but `./quantlib-test-suite -- --date=today` works just fine even with the parallel unit test runner.

Reply:
That's weird.  The test executes correctly on its own with the default runner, and it starts with the default date set.

Oh well — does it work if you move up [the definition of `today`](https://github.com/lballabio/QuantLib/blob/master/test-suite/piecewiseyieldcurve.cpp#L1170) and set it as the evaluation date here rather than [later](https://github.com/lballabio/QuantLib/blob/master/test-suite/piecewiseyieldcurve.cpp#L1173)?

Reply:
Unfortunately moving the statements does not help. This test does not execute on its own with the default runner on my system (linux, gcc 9.4) either. On my machine it is not an issue with the parallel test runner it just happens to come into the light because the execution order of the tests is not fixed.

Reply:
What's the error when it fails?

Reply:
Hi Luigi, 

I'm using a standard Ubuntu 20.04 configuration and cmake with no changes to the github head version. Even though the issue does not show up on my windows build (VS 2022) there might by a problem here and your intuition was correct, my original change is sort of a red herring as it does not cure the root cause. The observer notification when changing the evaluation date does not work correctly in my set-up for this test case and - IMHO by chance - the issue does not show up under Windows. 

When changing the evaluation date, the update method of SwapRateHelper creates a new[ Swap](https://github.com/lballabio/QuantLib/blob/master/ql/termstructures/yield/ratehelpers.cpp#L821), which creates new FloatingRatesCoupon. These floating rate coupons register with the [evaluation date](https://github.com/lballabio/QuantLib/blob/master/ql/cashflows/floatingratecoupon.cpp#L57) during construction, hence we are changing the set of observers (erase and insert) on the evaluation date [while iterating over the unordered_set.](https://github.com/lballabio/QuantLib/blob/master/ql/patterns/observable.cpp#L64) This does not work correctly on the Linux build, symptom being "lost notifications" or "pure virtual function call". Maybe going back to set is fixing the [issue](https://stackoverflow.com/questions/6438086/iterator-invalidation-rules-for-c-containers)(?), see Insertion on unordered_set when rehashing occurs.

Reply:
If invalidated iterators are being dereferenced then I'm surprised that the AddressSanitizer CI job doesn't catch it, but I suppose it's possible.

Even if changing `std::unordered_set` to `std::set` prevents iterator invalidation, new observers added while iterating could be added either before or after the current iterator depending on the sorting, which doesn't sound too desirable.

More generally, is changing the set of observers while iterating something that should be allowed? If so, should it be allowed while iterating, or should the library take care of queuing all pending changes to the set of observers until the current round of notifications is done?

Reply:
The latter one would be to change [here](https://github.com/lballabio/QuantLib/blob/master/ql/patterns/observable.cpp#L64) into
```
            for (auto* observer: std::vector<set_type::key_type>(
                    observers_.begin(), observers_.end())) {
                try {
                    if (observers_.count(observer))
                        observer->update();
                }
```
In this case newly created observers while looping over all observers would never be called but it might come with a performance hit. Replacing **unordered_set** with **set** is IMHO valid if we accept that newly created observers **might** be called upon the loop over all observers.

Reply:
Or we can postpone the creation of the swap after the loop is done.  I'll give it a try.

Reply:
No, I did give it half a try and it would be a mess of mutable data members and boolean flags.

I'd switch to std::set.  Technically it's not a backward-compatible change, but it's a necessary one and we prepared for that already by deprecating public usage of `unordered_set` in release 1.26.  We'll be removing it earlier than expected.  It doesn't look like the kind of typedef that people would actually use anyway.

Reply:
And we're iterating over the set, not looking up elements, so we're probably not losing performance.

Reply:
> Replacing unordered_set with set is IMHO valid if we accept that newly created observers might be called during the loop over all observers.

One other option, if we are going to change the data structure anyway, would be to use a list and just add the new observer at the end. This would address the iterator invalidation issue as well as the ordering issue. The cost would be checking for whether a new observer has already been added to the list at the time it’s added.

Reply:
True, but I'd keep it as simple as possible.  The set does the check for you.

Reply:
Actually I think we do not need to change unordered_set in Observer, keeping the change backward compatible. Only Observable needs to change. I've also added a test case to make the problem more obvious (hmm.. sort off).

Reply:
Thanks!